File: src/main/java/com/bajdcc/LALR1/interpret/module/ModuleFile.java
package com.bajdcc.LALR1.interpret.module;

import com.bajdcc.LALR1.grammar.Grammar;
import com.bajdcc.LALR1.grammar.runtime.*;
import com.bajdcc.util.ResourceLoader;

import java.math.BigInteger;
import java.util.List;

/**
 * 【模块】文件模块
 *
 * @author bajdcc
 */
public class ModuleFile implements IInterpreterModule {

	private static ModuleFile instance = new ModuleFile();
	private RuntimeCodePage runtimeCodePage;

	public static ModuleFile getInstance() {
		return instance;
	}

	@Override
	public String getModuleName() {
		return "sys.file";
	}

	@Override
	public String getModuleCode() {
		return ResourceLoader.load(getClass());
	}

	@Override
	public RuntimeCodePage getCodePage() throws Exception {
		if (runtimeCodePage != null)
			return runtimeCodePage;

		String base = ResourceLoader.load(getClass());

		Grammar grammar = new Grammar(base);
		RuntimeCodePage page = grammar.getCodePage();
		IRuntimeDebugInfo info = page.getInfo();
		buildMethod(info);

		return runtimeCodePage = page;
	}

	private void buildMethod(IRuntimeDebugInfo info) {
		info.addExternalFunc("g_create_file_internal", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "创建文件";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[]{RuntimeObjectType.kString, RuntimeObjectType.kInt, RuntimeObjectType.kString};
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) {
				String name = args.get(0).getObj().toString();
				int mode = ((BigInteger) args.get(1).getObj()).intValue();
				String encoding = args.get(2).getObj().toString();
				return new RuntimeObject(status.getService().getFileService().create(name, mode, encoding));
			}
		});
		info.addExternalFunc("g_destroy_file_internal", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "关闭文件";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[]{RuntimeObjectType.kPtr};
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) {
				int handle = (int) args.get(0).getObj();
				return new RuntimeObject(status.getService().getFileService().destroy(handle));
			}
		});
		info.addExternalFunc("g_write_file_internal", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "写文件";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[]{RuntimeObjectType.kPtr, RuntimeObjectType.kChar};
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) {
				int handle = (int) args.get(0).getObj();
				char ch = (char) args.get(1).getObj();
				return new RuntimeObject(status.getService().getFileService().write(handle, ch));
			}
		});
		info.addExternalFunc("g_write_file_string_internal", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "写文件（字串）";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[]{RuntimeObjectType.kPtr, RuntimeObjectType.kString};
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) {
				int handle = (int) args.get(0).getObj();
				String str = (String) args.get(1).getObj();
				return new RuntimeObject(status.getService().getFileService().writeString(handle, str));
			}
		});
		info.addExternalFunc("g_read_file_internal", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "读文件";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[]{RuntimeObjectType.kPtr};
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) {
				int handle = (int) args.get(0).getObj();
				int ch = status.getService().getFileService().read(handle);
				if (ch == -1)
					return null;
				return new RuntimeObject((char) ch);
			}
		});
		info.addExternalFunc("g_read_file_string_internal", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "读文件（行）";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[]{RuntimeObjectType.kPtr};
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) {
				int handle = (int) args.get(0).getObj();
				String str = status.getService().getFileService().readString(handle);
				return new RuntimeObject(str);
			}
		});
		info.addExternalFunc("g_query_file", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "判断文件是否存在";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[]{RuntimeObjectType.kString};
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) {
				String filename = String.valueOf(args.get(0).getObj());
				return new RuntimeObject(status.getService().getFileService().exists(filename));
			}
		});
		info.addExternalFunc("g_read_file_vfs_utf8", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "读文件（一次性），限VFS";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[]{RuntimeObjectType.kString};
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) {
				String filename = String.valueOf(args.get(0).getObj());
				return new RuntimeObject(status.getService().getFileService().readAll(filename));
			}
		});
	}
}

File: src/main/java/com/bajdcc/LALR1/interpret/module/ModuleFile.kt
package com.bajdcc.LALR1.interpret.module

import com.bajdcc.LALR1.grammar.Grammar
import com.bajdcc.LALR1.grammar.runtime.*
import com.bajdcc.util.ResourceLoader

/**
 * 【模块】文件模块
 *
 * @author bajdcc
 */
class ModuleFile : IInterpreterModule {
    private var runtimeCodePage: RuntimeCodePage? = null

    override val moduleName: String
        get() = "sys.file"

    override val moduleCode: String
        get() = ResourceLoader.load(javaClass)

    override val codePage: RuntimeCodePage
        @Throws(Exception::class)
        get() {
            if (runtimeCodePage != null)
                return runtimeCodePage!!

            val base = ResourceLoader.load(javaClass)

            val grammar = Grammar(base)
            val page = grammar.codePage
            val info = page.info
            buildMethod(info)

            runtimeCodePage = page
            return page
        }

    private fun buildMethod(info: IRuntimeDebugInfo) {
        info.addExternalFunc("g_create_file_internal",
                RuntimeDebugExec("创建文件", arrayOf(RuntimeObjectType.kString, RuntimeObjectType.kInt, RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val name = args[0].string
                        val mode = args[1].int
                        val encoding = args[2].string
                        RuntimeObject(status.service.fileService.create(name, mode, encoding, status.page))
                    }
                })
        info.addExternalFunc("g_destroy_file_internal",
                RuntimeDebugExec("关闭文件", arrayOf(RuntimeObjectType.kPtr))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val handle = args[0].int
                        RuntimeObject(status.service.fileService.destroy(handle))
                    }
                })
        info.addExternalFunc("g_write_file_internal",
                RuntimeDebugExec("写文件", arrayOf(RuntimeObjectType.kPtr, RuntimeObjectType.kChar))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val handle = args[0].int
                        val ch = args[1].char
                        RuntimeObject(status.service.fileService.write(handle, ch))
                    }
                })
        info.addExternalFunc("g_write_file_string_internal",
                RuntimeDebugExec("写文件（字串）", arrayOf(RuntimeObjectType.kPtr, RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val handle = args[0].int
                        val str = args[1].string
                        RuntimeObject(status.service.fileService.writeString(handle, str))
                    }
                })
        info.addExternalFunc("g_read_file_internal",
                RuntimeDebugExec("读文件", arrayOf(RuntimeObjectType.kPtr))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val handle = args[0].int
                        val ch = status.service.fileService.read(handle)
                        if (ch == -1) null else RuntimeObject(ch.toChar())
                    }
                })
        info.addExternalFunc("g_read_file_string_internal",
                RuntimeDebugExec("读文件（行）", arrayOf(RuntimeObjectType.kPtr))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val handle = args[0].int
                        val str = status.service.fileService.readString(handle)
                        RuntimeObject(str)
                    }
                })
        info.addExternalFunc("g_query_file",
                RuntimeDebugExec("判断文件是否存在", arrayOf(RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val filename = args[0].string
                        RuntimeObject(status.service.fileService.exists(filename))
                    }
                })
        info.addExternalFunc("g_read_file_vfs_utf8",
                RuntimeDebugExec("读文件（一次性），限VFS", arrayOf(RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val filename = args[0].string
                        RuntimeObject(status.service.fileService.readAll(filename))
                    }
                })
    }

    companion object {

        val instance = ModuleFile()
    }
}
